addons/sale/models/sale.py

```diff
@@ -1058,12 +1055,13 @@ def _prepare_invoice_line(self, qty):
         self.ensure_one()
         res = {}
         account = self.product_id.property_account_income_id or self.product_id.categ_id.property_account_income_categ_id
-        if not account:
+
+        if not account and self.product_id:
             raise UserError(_('Please define income account for this product: "%s" (id:%d) - or for its category: "%s".') %
                 (self.product_id.name, self.product_id.id, self.product_id.categ_id.name))
 
         fpos = self.order_id.fiscal_position_id or self.order_id.partner_id.property_account_position_id
-        if fpos:
+        if fpos and account:
```

Comment:
I chose account in this case because it is was is used inside the check. Even if they are related we can't care about product_id being set or not for the following logic, just that there is an account.